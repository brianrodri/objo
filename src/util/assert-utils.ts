import { attempt, isError } from "lodash";

/**
 * Asserts that all items in an array pass the given assertion.
 * Throws an aggregate error with _all_ failures, rather than just the first one encountered.
 * @typeParam T - the type of each input item.
 * @typeParam Constrained - the constrained type of each input item.
 * @param items - The array of items to inspect.
 * @param assertion - Used to assert that an item is constrained.
 * @param message - Optional message to include if an assertion fails.
 * @throws aggregate error with items that failed the assertion.
 */
export function assertEach<T, Constrained extends T>(
    items: T[],
    assertion: (item: T) => asserts item is Constrained,
    message?: string,
): asserts items is Constrained[] {
    const failures = items
        .map((item) => attempt(() => assertion(item)))
        .map((error, index) => (error ? new Error(`invalid index ${index}: ${error.message}`) : null))
        .filter(isError);
    if (failures.length > 0) {
        const messageLines = [
            message ?? `assertion failed on ${failures.length} item(s)`,
            ...failures.map((error) => error.message.replace(/^/gm, "\t")),
        ];
        throw new AggregateError(failures, messageLines.join("\n"));
    }
}
