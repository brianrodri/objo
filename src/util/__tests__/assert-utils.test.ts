import assert from "assert";
import { isNumber } from "lodash";
import { describe, expect, expectTypeOf, it } from "vitest";

import { assertEachWith } from "../assert-utils";

/** @param x - value that must be a number. */
function assertNumber(x: unknown): asserts x is number {
    assert(isNumber(x), `${JSON.stringify(x)} is not a number`);
}

describe(`${assertEachWith.name}`, () => {
    describe.each(["user-provided message", undefined])("with message=%j", (message) => {
        it("should accept valid values", () => {
            const items: (number | string)[] = [1, 2, 3, 4, 5];
            expectTypeOf(items).toEqualTypeOf<(number | string)[]>();
            assertEachWith(items, assertNumber, message);
            expectTypeOf(items).toEqualTypeOf<number[]>();
        });

        it("should reject invalid values", () => {
            const items: (number | string)[] = [1, "2", 3, "4", 5];
            expectTypeOf(items).toEqualTypeOf<(number | string)[]>();
            expect(() => assertEachWith(items, assertNumber, message)).toThrowErrorMatchingSnapshot();
            expectTypeOf(items).not.toEqualTypeOf<number[]>();
        });
    });
});
