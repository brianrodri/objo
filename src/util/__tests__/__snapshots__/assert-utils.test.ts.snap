// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`assertEachWith > with message="user-provided message" > should reject invalid values 1`] = `
[AggregateError: user-provided message
	invalid index 1: "2" is not a number
	invalid index 3: "4" is not a number]
`;

exports[`assertEachWith > with message=undefined > should reject invalid values 1`] = `
[AggregateError: assertion failed on 2 item(s)
	invalid index 1: "2" is not a number
	invalid index 3: "4" is not a number]
`;
